File: it/src/test/scala/org/zalando/nakadi/client/scala/SimpleEventTest.scala
package org.zalando.nakadi.client.scala

import org.scalatest.BeforeAndAfterAll
import org.scalatest.Matchers
import org.scalatest.WordSpec
import org.zalando.nakadi.client.scala.model.Cursor
import org.zalando.nakadi.client.scala.model.EventEnrichmentStrategy
import org.zalando.nakadi.client.scala.model.PartitionStrategy
import org.zalando.nakadi.client.scala.model.EventTypeStatistics
import org.zalando.nakadi.client.scala.model.PartitionStrategyType
import org.zalando.nakadi.client.scala.model.ScalaJacksonJsonMarshaller
import org.zalando.nakadi.client.scala.test.factory.EventIntegrationHelper
import org.zalando.nakadi.client.scala.test.factory.events.MySimpleEvent
import org.zalando.nakadi.client.scala.test.factory.events.SimpleEventListener
import org.zalando.nakadi.client.utils.ClientBuilder
import org.scalatest.BeforeAndAfter
import org.slf4j.LoggerFactory

class SimpleEventTest extends WordSpec with Matchers with BeforeAndAfter with BeforeAndAfterAll {
  import ClientFactory._
  import ScalaJacksonJsonMarshaller._
  import MySimpleEvent._

  private val client = ClientFactory.buildScalaClient()
  private var itEvent:EventIntegrationHelper = null
  
  val nrOfEvents = 45

  after{
    
    itEvent.deleteEventType()
  }
  override def afterAll {
        client.stop()
  }

  "404_should be handled graciously" in {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"SimpleEventIntegrationTest-Handle-404-Graciously"
    }
    itEvent = new EventIntegrationHelper(eventGenerator, client)
    itEvent.createEventType() shouldBe true

    itEvent.getEventType("non-existing-event-type-name") match {
      case Some(_) => fail("Should not fail, because eventType was not created yet!!")
      case None    =>
    }
  }

  "Validate Published nr of SimpleEvents" in {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"SimpleEventIntegrationTest-Validate-Published-Events-$nrOfEvents"
    }
    itEvent = new EventIntegrationHelper(eventGenerator, client)
    val cursor = Some(Cursor("0", "BEGIN"))
    val listener = new SimpleEventListener()

    itEvent.createEventType() shouldBe true
    Thread.sleep(2000)
    val events = itEvent.publishEvents(nrOfEvents)
    client.subscribe(eventGenerator.eventTypeName, StreamParameters(cursor = cursor), listener)
    val receivedEvents = listener.waitToReceive(nrOfEvents)
    receivedEvents.size shouldBe events.size
    receivedEvents shouldBe events
  }

  "Multiple events listeners must work in parallel" in {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"SimpleEventIntegrationTest-Multiple-listeners-in-parallel-$nrOfEvents"
    }
    itEvent = new EventIntegrationHelper(eventGenerator, client)
    val cursor = Some(Cursor("0", "BEGIN"))
    itEvent.createEventType() shouldBe true
    val events = itEvent.publishEvents(nrOfEvents)
    val listener = new SimpleEventListener()
    val listener2 = new SimpleEventListener()
    val listener3 = new SimpleEventListener()

    client.subscribe(eventGenerator.eventTypeName, StreamParameters(cursor = cursor), listener)
    client.subscribe(eventGenerator.eventTypeName, StreamParameters(cursor = cursor), listener2)

    val receivedEvents = listener.waitToReceive(nrOfEvents)
    receivedEvents.size shouldBe events.size
    receivedEvents shouldBe events

    val receivedEvents2 = listener2.waitToReceive(nrOfEvents)
    receivedEvents2.size shouldBe events.size
    receivedEvents2 shouldBe events

    client.subscribe(eventGenerator.eventTypeName, StreamParameters(cursor = cursor), listener)
    val receivedEvents3 = listener.waitToReceive(nrOfEvents * 2)
    receivedEvents3.size shouldBe (nrOfEvents * 2)

  }

  "An unsubscribed listener should not receive any events" in {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"SimpleEventIntegrationTest-Multiple-listeners-in-parallel-$nrOfEvents"
    }
    itEvent = new EventIntegrationHelper(eventGenerator, client)
    val cursor = Some(Cursor("0", "BEGIN"))
    itEvent.createEventType() shouldBe true
    val listener = new SimpleEventListener()
    client.subscribe(eventGenerator.eventTypeName, StreamParameters(cursor = cursor), listener)
    client.unsubscribe(eventGenerator.eventTypeName, Option("0"), listener)
    itEvent.publishEvents(nrOfEvents)
    Thread.sleep(5000)
    listener.receivedEvents.size shouldBe 0
  }
  "Validate created EventType" in {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"SimpleEventIntegrationTest-Validate-Created-EventType"
    }
    itEvent = new EventIntegrationHelper(eventGenerator, client)

    itEvent.createEventType() shouldBe true

    val optionalOfCreatedEventType = itEvent.getEventType()

    optionalOfCreatedEventType.isDefined shouldBe true

    val Some(eventType) = optionalOfCreatedEventType
    eventType.category shouldBe itEvent.eventType.category
    eventType.dataKeyFields shouldBe null
    eventType.name shouldBe itEvent.eventType.name
    eventType.owningApplication shouldBe itEvent.eventType.owningApplication
    eventType.partitionStrategy shouldBe itEvent.eventType.partitionStrategy
    eventType.schema shouldBe itEvent.eventType.schema
    eventType.statistics shouldBe itEvent.eventType.statistics
    eventType.enrichmentStrategies shouldBe itEvent.eventType.enrichmentStrategies
    eventType.partitionKeyFields shouldBe itEvent.eventType.partitionKeyFields
  }
  "Validate created EventType wiht Statistics" in {
      val eventGenerator = new DefaultMySimpleEventGenerator() {
          def eventTypeId = s"SimpleEventIntegrationTest-Validate-Created-EventType"
          override def statistics: Option[EventTypeStatistics] = Option(EventTypeStatistics(2400, 20240, 4, 4))
      }
      itEvent = new EventIntegrationHelper(eventGenerator, client)
      
      itEvent.createEventType() shouldBe true
      
      Thread.sleep(5000)
      
      val optionalOfCreatedEventType = itEvent.getEventType()
      
      optionalOfCreatedEventType.isDefined shouldBe true
      
      val Some(eventType) = optionalOfCreatedEventType
      eventType.category shouldBe itEvent.eventType.category
      eventType.dataKeyFields shouldBe null
      eventType.name shouldBe itEvent.eventType.name
      eventType.owningApplication shouldBe itEvent.eventType.owningApplication
      eventType.partitionStrategy shouldBe itEvent.eventType.partitionStrategy
      eventType.schema shouldBe itEvent.eventType.schema
      eventType.statistics shouldBe itEvent.eventType.statistics
      eventType.enrichmentStrategies shouldBe itEvent.eventType.enrichmentStrategies
      eventType.partitionKeyFields shouldBe itEvent.eventType.partitionKeyFields
      eventType.statistics should not be null 
      eventType.statistics.get.messageSize shouldBe itEvent.eventType.statistics.get.messageSize
      eventType.statistics.get.messagesPerMinute shouldBe itEvent.eventType.statistics.get.messagesPerMinute
      eventType.statistics.get.readParallelism shouldBe itEvent.eventType.statistics.get.readParallelism
      eventType.statistics.get.writeParallelism shouldBe itEvent.eventType.statistics.get.writeParallelism
      
  }

  "Validate nr of partitions after Creation of EventType" in {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"SimpleEventIntegrationTest-Validate-nr-of-partitions"
    }
    itEvent = new EventIntegrationHelper(eventGenerator, client)
    itEvent.createEventType() shouldBe true
    itEvent.getNumberOfPartitions() shouldBe 1
  }

  "Receive partition-strategies successfully" ignore {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"SimpleEventIntegrationTest-Receive-partition-strategies-successfully"
    }
    itEvent = new EventIntegrationHelper(eventGenerator, client)
    val result = itEvent.getPartitionStrategies()
    result.size shouldBe 3 //NOT IMPLEMENTED
    result should contain(PartitionStrategy.HASH)
    result should contain(PartitionStrategy.RANDOM)
    result should contain(PartitionStrategy.USER_DEFINED)
  }

  "Receive enrichment-strategies successfully" ignore {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"SimpleEventIntegrationTest-Receive-enrichment-strategies-successfully"
    }
    itEvent = new EventIntegrationHelper(eventGenerator, client)
    val result = itEvent.getEnrichmentStrategies()
    result.size shouldBe 1
    result should contain(EventEnrichmentStrategy.METADATA)
  }
  
}

File: it/src/test/java/org/zalando/client/java/SimpleEventTest.java
package org.zalando.client.java;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.ExecutionException;

import org.junit.After;
import org.junit.Test;
import org.zalando.nakadi.client.java.Client;
import org.zalando.nakadi.client.java.StreamParameters;
import org.zalando.nakadi.client.java.enumerator.PartitionStrategy;
import org.zalando.nakadi.client.java.model.Cursor;
import org.zalando.nakadi.client.java.model.Event;
import org.zalando.nakadi.client.java.model.EventStreamBatch;
import org.zalando.nakadi.client.java.model.EventType;
import org.zalando.nakadi.client.java.model.EventTypeStatistics;
import org.zalando.nakadi.client.java.test.event.generator.EventGenerator;
import org.zalando.nakadi.client.java.test.event.generator.EventIntegrationHelper;
import org.zalando.nakadi.client.java.test.event.simple.MySimpleEvent;
import org.zalando.nakadi.client.java.test.event.simple.MySimpleEventGenerator;
import org.zalando.nakadi.client.java.test.event.simple.SimpleEventListener;
import org.zalando.nakadi.client.scala.ClientFactory;

import com.fasterxml.jackson.core.type.TypeReference;

public class SimpleEventTest {

    private Client client = ClientFactory.buildJavaClient();
    private Integer nrOfEvents = 45;

    private TypeReference<EventStreamBatch<MySimpleEvent>> typeRef = new TypeReference<EventStreamBatch<MySimpleEvent>>() {
    };

    @After
    public void shutdown() throws InterruptedException, ExecutionException {
        client.stop();
    }

    @Test
    public void handle404Graciously() throws InterruptedException, ExecutionException {

        EventGenerator gen = new MySimpleEventGenerator()//
                .withEventTypeId("SimpleEventTest-handle404Graciously")//
                .build();
        EventIntegrationHelper it = new EventIntegrationHelper(gen, client);
        assertTrue(it.createEventType());
        assertTrue(it.getEventType().isPresent());
        Optional<EventType> result = client.getEventType("none-existing-event-type-name").get();
        assertEquals(result.isPresent(), false);
    }

    @Test
    public void validatePublishedNrOfEvents() throws InterruptedException, ExecutionException {
        SimpleEventListener listener = new SimpleEventListener();
        EventGenerator gen = new MySimpleEventGenerator()//
                .withEventTypeId("SimpleEventTest-validatePublishedNrOfEvents").build();
        EventIntegrationHelper it = new EventIntegrationHelper(gen, client);
        assertTrue("EventType should be created", it.createEventType());
        Thread.sleep(5000);// Creation can take time.
        Optional<EventType> eventTypeOpt = it.getEventType();
        assertTrue("Did not return the eventType", eventTypeOpt.isPresent());
        List<Event> createdEvents = it.publishEvents(nrOfEvents);
        Optional<Cursor> cursor = Optional.of(new Cursor("0", "BEGIN"));
        Optional<Integer> batchLimit = Optional.empty();
        Optional<Integer> streamLimit = Optional.empty();
        Optional<Integer> batchFlushTimeout = Optional.empty();
        Optional<Integer> streamTimeout = Optional.empty();
        Optional<Integer> streamKeepAliveLimit = Optional.empty();
        Optional<String> flowId = Optional.empty();
        StreamParameters parameters = new StreamParameters(cursor, //
                batchLimit, //
                streamLimit, //
                batchFlushTimeout, //
                streamTimeout, //
                streamKeepAliveLimit, //
                flowId);
        client.subscribe(it.getGen().getEventTypeName(), parameters, listener, typeRef);
        List<MySimpleEvent> receivedEvents = listener.waitToReceive(nrOfEvents);
        assertEquals("Created & Received events differ in number", createdEvents.size(), receivedEvents.size());
    }

    @Test
    public void unsubscribedListenerShouldNotReceiveAnyEvents() throws InterruptedException, ExecutionException {
        SimpleEventListener listener = new SimpleEventListener();
        EventGenerator gen = new MySimpleEventGenerator()//
                .withEventTypeId("SimpleEventTest-validatePublishedNrOfEvents").build();
        EventIntegrationHelper it = new EventIntegrationHelper(gen, client);
        assertTrue("EventType should be created", it.createEventType());
        Thread.sleep(1000);// Creation can take time.
        Optional<EventType> eventTypeOpt = it.getEventType();
        assertTrue("Did not return the eventType", eventTypeOpt.isPresent());

        Optional<Cursor> cursor = Optional.of(new Cursor("0", "BEGIN"));
        Optional<Integer> batchLimit = Optional.empty();
        Optional<Integer> streamLimit = Optional.empty();
        Optional<Integer> batchFlushTimeout = Optional.empty();
        Optional<Integer> streamTimeout = Optional.empty();
        Optional<Integer> streamKeepAliveLimit = Optional.empty();
        Optional<String> flowId = Optional.empty();
        StreamParameters parameters = new StreamParameters(cursor, //
                batchLimit, //
                streamLimit, //
                batchFlushTimeout, //
                streamTimeout, //
                streamKeepAliveLimit, //
                flowId);
        client.subscribe(it.getGen().getEventTypeName(), parameters, listener, typeRef);
        client.unsubscribe(it.getGen().getEventTypeName(), Optional.of("0"), listener);
        it.publishEvents(nrOfEvents);
        Thread.sleep(5000);
        assertEquals("Unsubscribed listener must not receive events", 0, listener.getReceivedEvents().size());
    }

    @Test
    public void validateCreatedEventType() throws InterruptedException {
        EventGenerator gen = new MySimpleEventGenerator()//
                .withEventTypeId("SimpleEventTest-validateCreatedEventType").build();
        EventIntegrationHelper it = new EventIntegrationHelper(gen, client);
        assertTrue("EventType should be created", it.createEventType());
        Thread.sleep(1000);// Creation can take time.
        Optional<EventType> eventTypeOpt = it.getEventType();
        assertTrue("Did not return the eventType", eventTypeOpt.isPresent());

        EventType originalEventType = it.getEventType().get();
        EventType eventType = eventTypeOpt.get();
        assertEquals(eventType.getCategory(), originalEventType.getCategory());
        assertEquals(eventType.getDataKeyFields(), originalEventType.getDataKeyFields());
        assertEquals(eventType.getName(), originalEventType.getName());
        assertEquals(eventType.getOwningApplication(), originalEventType.getOwningApplication());
        assertEquals(eventType.getPartitionKeyFields(), originalEventType.getPartitionKeyFields());
        assertEquals(eventType.getPartitionStrategy(), originalEventType.getPartitionStrategy());
        assertEquals(eventType.getSchema().getSchema(), originalEventType.getSchema().getSchema());
        assertEquals(eventType.getSchema().getType(), originalEventType.getSchema().getType());
        assertNull(eventType.getStatistics());
    }
    @Test
    public void validateCreatedEventTypeWithStatistics() throws InterruptedException {
        Integer messagesPerMinute = 2400;
        Integer messageSize = 20240;
        Integer readParallelism = 8;
        Integer writeParallelism = 7;
        EventGenerator gen = new MySimpleEventGenerator(){
            @Override
          public EventTypeStatistics getStatistics() {
            return new EventTypeStatistics(messagesPerMinute, messageSize, readParallelism, writeParallelism);
          }
        }
        .withEventTypeId("SimpleEventTest-validateCreatedEventType").build();
        EventIntegrationHelper it = new EventIntegrationHelper(gen, client);
        assertTrue("EventType should be created", it.createEventType());
        Thread.sleep(1000);// Creation can take time.
        Optional<EventType> eventTypeOpt = it.getEventType();
        assertTrue("Did not return the eventType", eventTypeOpt.isPresent());
        
        EventType originalEventType = it.getEventType().get();
        EventType eventType = eventTypeOpt.get();
        assertEquals(eventType.getCategory(), originalEventType.getCategory());
        assertEquals(eventType.getDataKeyFields(), originalEventType.getDataKeyFields());
        assertEquals(eventType.getName(), originalEventType.getName());
        assertEquals(eventType.getOwningApplication(), originalEventType.getOwningApplication());
        assertEquals(eventType.getPartitionKeyFields(), originalEventType.getPartitionKeyFields());
        assertEquals(eventType.getPartitionStrategy(), originalEventType.getPartitionStrategy());
        assertEquals(eventType.getSchema().getSchema(), originalEventType.getSchema().getSchema());
        assertEquals(eventType.getSchema().getType(), originalEventType.getSchema().getType());
        assertNotNull(eventType.getStatistics());
        assertEquals(messageSize, originalEventType.getStatistics().getMessageSize());
        assertEquals(messagesPerMinute, originalEventType.getStatistics().getMessagesPerMinute());
        assertEquals(readParallelism, originalEventType.getStatistics().getReadParallelism());
        assertEquals(writeParallelism, originalEventType.getStatistics().getWriteParallelism());
        assertEquals(eventType.getStatistics().getMessageSize(), originalEventType.getStatistics().getMessageSize());
        assertEquals(eventType.getStatistics().getMessagesPerMinute(), originalEventType.getStatistics().getMessagesPerMinute());
        assertEquals(eventType.getStatistics().getReadParallelism(), originalEventType.getStatistics().getReadParallelism());
        assertEquals(eventType.getStatistics().getWriteParallelism(), originalEventType.getStatistics().getWriteParallelism());
    }

    @Test
    public void validateNrOfPartition() throws InterruptedException, ExecutionException {
        EventGenerator gen = new MySimpleEventGenerator()//
                .withEventTypeId("SimpleEventTest-validateNrOfPartition").build();
        EventIntegrationHelper it = new EventIntegrationHelper(gen, client);
        assertTrue("EventType should be created", it.createEventType());
        assertEquals(it.getNumberOfPartitions(), Integer.valueOf(1));

    }

    @Test
    public void receivePartitionStrategies() throws InterruptedException, ExecutionException {
        EventGenerator gen = new MySimpleEventGenerator()//
                .withEventTypeId("SimpleEventTest-receivePartitionStrategies").build();
        EventIntegrationHelper it = new EventIntegrationHelper(gen, client);
        List<PartitionStrategy> strategies = it.getPartitionStrategies();
        assertEquals(strategies.size(), 3);
        for (PartitionStrategy ps : strategies) {
            assertTrue("Did not find PartitionStrategy:" + ps.name(), PartitionStrategy.withName(ps.name()).isPresent());

        }
    }

}
